asapdiscovery-alchemy/asapdiscovery/alchemy/utils.py

```diff
@@ -175,3 +176,30 @@ def collect_results(
         )
 
         return network_with_results
+
+    def collect_errors(
+        self,
+        planned_network: FreeEnergyCalculationNetwork,
+        with_traceback: bool = False,
+    ) -> dict[str, dict[str, str]]:
+        """
+        Collect errors from failed tasks.
+
+        Args:
+            planned_network: Network to get failed tasks from.
+            with_traceback: Output the complete traceback for the failed tasks.
+
+        Returns:
+            Nested dictionary with task key as value and a dictionary with errors and tracebacks as values.
+        """
+        network_key = planned_network.results.network_key
+        errored_tasks = self._client.get_network_tasks(network_key, status="error")
+
+        error_data = defaultdict(dict)
+        for task in errored_tasks:
+            for err_result in self._client.get_task_failures(task):
+                for failure in err_result.protocol_unit_failures:
+                    error_data[str(task.gufe_key)]["errors"] = failure.exception
+                    if with_traceback:
+                        error_data[str(task.gufe_key)]["traceback"] = failure.traceback
```

Comment:
These lines as written will only retain the last `protocol_unit_failure` exception/traceback for the last failed `ProtocolDAGResult` returned by `AlchemiscaleClient.get_task_failures`. This isn't necessarily complete information, since each `Task` can have many failed `ProtocolDAGResult`s (one for each attempt), and each of these could feature many failed units (eventually...with the way we currently run on `alchemiscale`, this isn't possible).

Reply:
Oh yes, you are right, let me think about it since that's definitely the general case we want to support. Good catch, thanks!

Reply:
One way is to just add another level to the nested object that has the protocol unit key, since a task can have many different failing units. I don't know if this dictionary is becoming too big and if so, would it be best to have our own custom object to store these results?